javascript/ql/src/semmle/javascript/frameworks/WebSocket.qll

```diff
@@ -0,0 +1,185 @@
+/**
+ * Provides classes for working with [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket).
+ *
+ * The model is based on the EventEmitter model, and there a therefore a
+ * data-flow step from where a WebSocket event is send to where the message
+ * is received.
+ *
+ * WebSockets include no concept of channels, therefore every client can send
+ * to every server (and vice versa).
+ */
+
+import javascript
+
+/**
+ * Gets the channel name used throughout this WebSocket model.
+ * WebSockets don't have a concept of channels, and therefore a singleton name is used.
+ * The name can be anything, as long as it is used consistently in this WebSocket model.
+ */
+private string channelName() { result = "message" }
+
+module ClientWebSocket {
+  /**
+   * A class that can be used to instantiate a WebSocket instance.
+   */
+  class SocketClass extends DataFlow::SourceNode {
+    boolean isNode;
+
+    SocketClass() {
+      this = DataFlow::globalVarRef("WebSocket") and isNode = false
+      or
+      this = DataFlow::moduleImport("ws") and isNode = true
+    }
+
+    /**
+     * Holds if this class an import of the "ws" module.
+     */
+    predicate isNode() { isNode = true }
+  }
+
+  /**
+   * A client WebSocket instance.
+   */
+  class ClientSocket extends EventEmitter::Range, DataFlow::SourceNode {
+    SocketClass socketClass;
+
+    ClientSocket() { this = socketClass.getAnInstantiation() }
+
+    /**
+     * Holds if this ClientSocket is created from the "ws" module.
+     *
+     * The predicate is used to differentiate where the behavior of the "ws" module differs from the native WebSocket in browsers.
+     */
+    predicate isNode() { socketClass.isNode() }
+  }
+
+  /**
+   * A message sent from a WebSocket client.
+   */
+  class SendNode extends EventDispatch::Range, DataFlow::CallNode {
+    override ClientSocket emitter;
+
+    SendNode() { this = emitter.getAMemberCall("send") }
+
+    override string getChannel() { result = channelName() }
+
+    override DataFlow::Node getSentItem(int i) { i = 0 and result = this.getArgument(0) }
+
+    override ServerWebSocket::ReceiveNode getAReceiver() { any() }
```

Comment:
The only possible way we have is based on the url, there is nothing else.     
On the server side we can only reasonably know the port ([like here](https://lgtm.com/projects/g/brettz9/webappfind/snapshot/6aed04f9356d613ae8c1143ce08002d42adee96f/files/native-app.js?sort=name&dir=ASC&mode=heatmap#xc1b337f94338b79:1)).  

On WebSocket clients we can also usually get the port, so I could try to treat the port like a channel.

Reply:
I'm also a little nervous about doing this without more data to back up the decision.

There may be other ways, such as matching parts of the message being sent. For example, if a message `{kind: "popup", url: x}` is sent, it could flow to a receiver in a branch where `msg.kind` has been checked to be `"popup"` - effectively recognizing ad-hoc channels. I have some ideas about how this could work, but again, without more data it's hard to know what works in practice.

Reply:
Could we design a experiment for *all* projects where we measure how many receivers each sender has? If a sender has >100 receivers in one project, then this implementation is definitely too imprecise...

Reply:
> Could we design a experiment for _all_ projects where we measure how many receivers each sender has? If a sender has >100 receivers in one project, then this implementation is definitely too imprecise...

https://lgtm.com/query/2587202964046394814/

Reply:
:+1: I will try this out using my newly aquired powers

Reply:
Let me know what you think of the extreme cases i sent you.

Reply:
> Let me know what you think of the extreme cases i sent you.

3 of them are just test-cases. So lots of short-lived clients and servers that use the same port.  

The last extreme case is 1 server that actually communicate with lots of clients (it's the result of a competition with lots of submissions).

Reply:
And there doesn't seem to be anything we can do to limit the amount of receivers for a given sender.  
The URL the WebSockets connect to is often a computed string based on some configuration.